backend/src/main/java/com/yohoard/notes/TagService.java

```diff
@@ -16,24 +15,22 @@ public TagService(TagRepository tagRepository) {
         this.tagRepository = tagRepository;
     }
 
-    public List<Tag> getOrCreate(List<String> names) {
-        List<Tag> existingTags = tagRepository.findByNameIn(names);
+    public List<Tag> getOrCreate(Set<String> names) {
```

Comment:
Chodziło mi raczej o coś takiego:

public List<Tag> getOrCreate(Set<String> names) {
        List<Tag> tags = tagRepository.findByNameIn(names);
        
        Set<Tag> newTags = tags.stream()
            .filter(tag -> !names.contains(tag.getName()))
            .map(tag -> new Tag(tag.getName(), "", 0))
            .collect(Collectors.toUnmodifiableSet());

        tagRepository.saveAll(newTags);

        tags.addAll(newTags);

        return tags;
    }

Wyciąganie nazw tagów do osobnej kolekcji było zbędne, skoro w parametrze są już one przekazane. Trzeba je tylko porównać z tym co już jest w bazie. No i przydałby się collector toUnmodifiableSet(), bo to i tak będzie niemutowalny set.

Ogólnie tu jest jeszcze miejsce na refaktor, ale to już najwyżej na później zostanie.